File: Assets/Scripts/Application/UIViews/UIBagItem.cs
#if !XLUA
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;
using TMPro;

namespace SkierFramework
{
    /// <summary>
    /// 背包列表项 C# 薄壳
    /// 继承 UILoopItem，由 UIScrollView 管理复用
    /// </summary>
    public class UIBagItem : UILoopItem
    {
        [ControlBinding] private TextMeshProUGUI TxtName;
        [ControlBinding] private TextMeshProUGUI TxtCount;
        [ControlBinding] private Image ImgIcon;
        [ControlBinding] private Image ImgQuality;
        [ControlBinding] private Button BtnItem;

        private BagItemData _data;

        public override void OnInit()
        {
            base.OnInit();
        }

        public override void OnAddListener()
        {
            base.OnAddListener();
            if (BtnItem != null)
            {
                BtnItem.AddClick(OnClickItem);
            }
        }

        protected override void OnUpdateData(IList dataList, int index, object userData)
        {
            if (index < 0 || index >= dataList.Count) return;

            _data = dataList[index] as BagItemData;
            if (_data == null) return;

            if (TxtName != null)
                TxtName.text = _data.name;

            if (TxtCount != null)
                TxtCount.text = _data.count > 1 ? _data.count.ToString() : "";

            // 品质颜色
            if (ImgQuality != null)
            {
                ImgQuality.color = GetQualityColor(_data.quality);
            }

            // 图标加载（通过Addressables异步加载）
            if (ImgIcon != null && !string.IsNullOrEmpty(_data.icon))
            {
                // 简化处理：实际项目中应通过 ResourceManager 异步加载图标
                // ResourceManager.Instance.LoadAssetAsync<Sprite>(iconPath, (sprite) => { ImgIcon.sprite = sprite; });
            }
        }

        public override void CheckSelect(int index)
        {
            // 可以在这里做选中高亮效果
        }

        private void OnClickItem()
        {
            if (_data == null) return;

            // 发事件给 Lua 层处理
#if XLUA
            var luaEnv = LuaEnvManager.Instance?.LuaEnv;
            if (luaEnv != null)
            {
                luaEnv.DoString(
                    $"EventBus:Emit('BagItemClicked', {_data.slotIndex}, {_data.itemId})",
                    "BagItemClick");
            }
#endif
        }

        private Color GetQualityColor(int quality)
        {
            switch (quality)
            {
                case 1: return Color.white;
                case 2: return Color.green;
                case 3: return new Color(0.3f, 0.5f, 1f);  // 蓝
                case 4: return new Color(0.7f, 0.3f, 0.9f); // 紫
                case 5: return new Color(1f, 0.6f, 0f);      // 橙
                default: return Color.gray;
            }
        }
    }
}
#endif

File: Assets/Lua/UI/Bag/UIBagItem.lua
---============================================================
--- UIBagItem.lua - 背包物品格子 Lua 控制器
---
--- 由通用 UILuaItem 转发生命周期
--- self.csItem 指向 C# 的 UILuaItem 实例
--- OnUpdateData 接收的 data 是一个 Lua table（来自 BagSystem:GetItemList）
---============================================================

local UIBagItem = {}

function UIBagItem:OnInit()
    print("[UIBagItem] OnInit")
end

function UIBagItem:OnAddListener()
    if self.BtnItem then
        self.BtnItem.onClick:AddListener(function() self:OnClick() end)
    end
end

function UIBagItem:OnUpdateData(data, index)
    if not data then return end

    self.slotIndex = data.slotIndex
    self.itemId = data.itemId

    if self.TxtName then
        self.TxtName.text = data.name or ""
    end
    if self.TxtCount then
        local countStr = ""
        if data.count and data.count > 1 then
            countStr = tostring(data.count)
        end
        self.TxtCount.text = countStr
    end
    -- 品质颜色
    if self.ImgQuality and data.quality then
        self.ImgQuality.color = self:GetQualityColor(data.quality)
    end
end

function UIBagItem:OnClick()
    if self.slotIndex and self.itemId then
        EventBus:Emit("BagItemClicked", self.slotIndex, self.itemId)
    end
end

function UIBagItem:GetQualityColor(quality)
    local CS = CS
    local Color = CS.UnityEngine.Color
    if quality == 1 then return Color.white
    elseif quality == 2 then return Color.green
    elseif quality == 3 then return CS.UnityEngine.Color(0.3, 0.5, 1, 1)
    elseif quality == 4 then return CS.UnityEngine.Color(0.7, 0.3, 0.9, 1)
    elseif quality == 5 then return CS.UnityEngine.Color(1, 0.6, 0, 1)
    else return Color.gray end
end

function UIBagItem:OnRelease()
    if self.BtnItem then
        self.BtnItem.onClick:RemoveAllListeners()
    end
end

return UIBagItem
